### PR TITLE
Add CommentUtil class for converting between Entity and Comment

### PR DESCRIPTION
**Pull request by n-wach**

### CODE REVIEW

File: src/main/java/com/googleinterns/zoomtube/utils/CommentUtil.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    Optional<Key> parent = Optional.ofNullable((Key) entity.getProperty(PARENT));
+    Date timestamp = (Date) entity.getProperty(TIMESTAMP);
+    User author = (User) entity.getProperty(AUTHOR);
+    String content = (String) entity.getProperty(CONTENT);
+    Date created = (Date) entity.getProperty(CREATED);
+    return Comment.create(key, lecture, parent, timestamp, author, content, created);
+  }
+
+  /**
+   * Creates and returns an entity for {@code comment}.
```
Comment (jaysaleh) [inline src/main/java/com/googleinterns/zoomtube/utils/CommentUtil.java:50]: Where is `comment` here? Maybe rephrase this

File: src/main/java/com/googleinterns/zoomtube/data/Comment.java [hunk 2]
```diff
@@ -24,34 +24,11 @@
...
-  public static final String ENTITY_KIND = "Comment";
-  public static final String PROP_LECTURE = "lecture";
-  public static final String PROP_PARENT = "parent";
-  public static final String PROP_TIMESTAMP = "timestamp";
-  public static final String PROP_AUTHOR = "author";
-  public static final String PROP_CONTENT = "content";
-  public static final String PROP_CREATED = "created";
-
-  private static Comment create(Key key, Key lecture, Optional<Key> parent, Date timestamp,
+  public static Comment create(Key key, Key lecture, Optional<Key> parent, Date timestamp,
```
Comment (amuamushu) [inline src/main/java/com/googleinterns/zoomtube/data/Comment.java:27]: Do you think you can rename `key` and `lecture` to be `commentKey` and `lectureKey` just so there won't be any confusion.

Comment (n-wach) [reply src/main/java/com/googleinterns/zoomtube/data/Comment.java]: Done!

File: src/main/java/com/googleinterns/zoomtube/utils/CommentUtil.java (hunk 1 above)
Comment (n-wach) [reply src/main/java/com/googleinterns/zoomtube/utils/CommentUtil.java]: Oops, copied this from somewhere, changed.

**Comment by n-wach** [comment]
Will wait until `lecture-entity-util` is merged into master

File: src/main/java/com/googleinterns/zoomtube/utils/CommentUtil.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.users.User;
+import com.googleinterns.zoomtube.data.Comment;
+import java.util.Date;
+import java.util.Optional;
+
+/** Provides methods to create Comment Entities and Comments. */
+public class CommentUtil {
```
Comment (f-camp) [inline src/main/java/com/googleinterns/zoomtube/utils/CommentUtil.java:25]: To enforce that this is a util class please make this final and add a private constructor. This will enforce it is not extendable and non-instantiable.

Comment (n-wach) [reply src/main/java/com/googleinterns/zoomtube/utils/CommentUtil.java:25]: Done!

File: src/main/java/com/googleinterns/zoomtube/utils/CommentUtil.java (hunk 1 above)
Comment (billie-cat) [reply src/main/java/com/googleinterns/zoomtube/utils/CommentUtil.java]: +1

File: src/test/java/com/googleinterns/zoomtube/utils/CommentUtilTest.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+  @Test
+  public void fromEntity_shouldReturnCommentFromEntity() throws IOException {
+    Key lecture = KeyFactory.createKey(LectureUtil.KIND, 12345);
+    Key parent = KeyFactory.createKey(CommentUtil.KIND, 67890);
+    Date timestamp = new Date(123);
+    User author = new User("test@example.com", "example.com");
+    String content = "Test content";
+    Date created = new Date();
+
```
Comment (billie-cat) [inline src/test/java/com/googleinterns/zoomtube/utils/CommentUtilTest.java:57]: please remove this new line.

File: src/main/java/com/googleinterns/zoomtube/utils/CommentUtil.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+  public static final String PARENT = "parent";
+  public static final String TIMESTAMP = "timestamp";
+  public static final String AUTHOR = "author";
+  public static final String CONTENT = "content";
+  public static final String CREATED = "created";
+
+  /**
+   * Returns a Comment using the properties of {@code entity}.
+   */
+  public static Comment fromEntity(Entity entity) {
```
Comment (billie-cat) [inline src/main/java/com/googleinterns/zoomtube/utils/CommentUtil.java:39]: nit: I would name this createComment because you are creating a Comment. Also this naming would be consistent with createEntity.

Comment (n-wach) [reply src/main/java/com/googleinterns/zoomtube/utils/CommentUtil.java:39]: Done

File: src/test/java/com/googleinterns/zoomtube/utils/CommentUtilTest.java (hunk 4 above)
Comment (n-wach) [reply src/test/java/com/googleinterns/zoomtube/utils/CommentUtilTest.java:57]: Done

File: src/main/java/com/googleinterns/zoomtube/servlets/DiscussionServlet.java [hunk 6]
```diff
@@ -62,34 +64,23 @@ public void init() throws ServletException {
...
-    commentEntity.setProperty(Comment.PROP_PARENT, parent);
-    // TODO: Add support for timestamped comments
-    commentEntity.setProperty(Comment.PROP_TIMESTAMP, new Date(0));
-    commentEntity.setProperty(Comment.PROP_AUTHOR, author);
-    commentEntity.setProperty(Comment.PROP_CONTENT, content);
-    // We set creation time to the time of the post request (now).
-    commentEntity.setProperty(Comment.PROP_CREATED, new Date(Clock.systemUTC().millis()));
-
-    datastore.put(commentEntity);
+    Entity comment = CommentUtil.createEntity(lecture, Optional.ofNullable(parent), new Date(0),
```
[the remark this answers is not in the document]
Comment (n-wach) [reply src/main/java/com/googleinterns/zoomtube/servlets/DiscussionServlet.java:81]: Changed

File: src/main/java/com/googleinterns/zoomtube/data/Comment.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (n-wach) [reply src/main/java/com/googleinterns/zoomtube/data/Comment.java]: I added a builder, but the caller still passes optional to the builder because Optional.ofNullable is simpler than an if statement

File: src/main/java/com/googleinterns/zoomtube/utils/CommentUtil.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+public final class CommentUtil {
+  public static final String KIND = "Comment";
+  public static final String LECTURE = "lecture";
+  public static final String PARENT = "parent";
+  public static final String TIMESTAMP = "timestamp";
+  public static final String AUTHOR = "author";
+  public static final String CONTENT = "content";
+  public static final String CREATED = "created";
+
+  private CommentUtil(){};
```
Comment (billie-cat) [inline src/main/java/com/googleinterns/zoomtube/utils/CommentUtil.java:34]:
```suggestion
  private CommentUtil() {};
```

Comment (billie-cat) [reply src/main/java/com/googleinterns/zoomtube/utils/CommentUtil.java:34]: nit: move this constructor at the bottom because it isn't important.

File: src/main/java/com/googleinterns/zoomtube/utils/CommentUtil.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+  /**
+   * Returns a Comment using the properties of {@code entity}.
```
Comment (billie-cat) [inline src/main/java/com/googleinterns/zoomtube/utils/CommentUtil.java:37]:
```suggestion
   * Creates and returns a Comment using the properties of {@code entity}.
```

File: src/test/java/com/googleinterns/zoomtube/utils/CommentUtilTest.java [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+    testServices.setUp();
+  }
+
+  @After
+  public void tearDown() {
+    testServices.tearDown();
+  }
+
+  @Test
+  public void fromEntity_shouldReturnCommentFromEntity() throws IOException {
```
Comment (billie-cat) [inline src/test/java/com/googleinterns/zoomtube/utils/CommentUtilTest.java:50]: please rename the test as it is not createComment method name

File: src/test/java/com/googleinterns/zoomtube/utils/CommentUtilTest.java [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+        lectureKey, Optional.of(parentKey), timestamp, author, content, dateNow);
+
+    assertThat(entity.getProperty(CommentUtil.LECTURE)).isEqualTo(lectureKey);
+    assertThat(entity.getProperty(CommentUtil.PARENT)).isEqualTo(parentKey);
+    assertThat(entity.getProperty(CommentUtil.TIMESTAMP)).isEqualTo(timestamp);
+    assertThat(entity.getProperty(CommentUtil.AUTHOR)).isEqualTo(author);
+    assertThat(entity.getProperty(CommentUtil.CONTENT)).isEqualTo(content);
+    assertThat(entity.getProperty(CommentUtil.CREATED)).isEqualTo(dateNow);
+  }
+}
```
Comment (billie-cat) [inline src/test/java/com/googleinterns/zoomtube/utils/CommentUtilTest.java:96]: please add another unit test for  creating an entity with no parent as that is an edge case.

File: src/main/java/com/googleinterns/zoomtube/data/Comment.java [hunk 11]
```diff
@@ -77,4 +48,22 @@ public static Comment fromEntity(Entity entity) {
...
+  public static Builder builder() {
+    return new AutoValue_Comment.Builder();
+  }
+
+  @AutoValue.Builder
+  public abstract static class Builder {
+    public abstract Builder setCommentKey(Key commentKey);
+    public abstract Builder setLectureKey(Key lectureKey);
+    public abstract Builder setParentKey(Key parentKey);
+    public abstract Builder setParentKey(Optional<Key> parentKey);
```
[the remark this answers is not in the document]
Comment (n-wach) [reply src/main/java/com/googleinterns/zoomtube/data/Comment.java:61]: Fixed!

File: src/main/java/com/googleinterns/zoomtube/utils/CommentUtil.java (hunk 7 above)
Comment (n-wach) [reply src/main/java/com/googleinterns/zoomtube/utils/CommentUtil.java:34]: The linter seems to remove the space automatically...

File: src/main/java/com/googleinterns/zoomtube/utils/CommentUtil.java (hunk 8 above)
Comment (n-wach) [reply src/main/java/com/googleinterns/zoomtube/utils/CommentUtil.java:37]: Changed

File: src/test/java/com/googleinterns/zoomtube/utils/CommentUtilTest.java (hunk 9 above)
Comment (n-wach) [reply src/test/java/com/googleinterns/zoomtube/utils/CommentUtilTest.java:50]: Done

File: src/test/java/com/googleinterns/zoomtube/utils/CommentUtilTest.java (hunk 10 above)
Comment (n-wach) [reply src/test/java/com/googleinterns/zoomtube/utils/CommentUtilTest.java:96]: Added

File: src/main/java/com/googleinterns/zoomtube/utils/CommentUtil.java (hunk 7 above)
Comment (n-wach) [reply src/main/java/com/googleinterns/zoomtube/utils/CommentUtil.java]: Moved constructor to bottom

File: src/main/java/com/googleinterns/zoomtube/servlets/DiscussionServlet.java [hunk 12]
```diff
@@ -62,34 +64,29 @@ public void init() throws ServletException {
...
-    commentEntity.setProperty(Comment.PROP_AUTHOR, author);
-    commentEntity.setProperty(Comment.PROP_CONTENT, content);
-    // We set creation time to the time of the post request (now).
-    commentEntity.setProperty(Comment.PROP_CREATED, new Date(Clock.systemUTC().millis()));
-
-    datastore.put(commentEntity);
+    final Entity comment;
+    if (parent == null) {
+      comment = CommentUtil.createEntityNoParent(
+          lecture, new Date(0), author, content, new Date(Clock.systemUTC().millis()));
```
Comment (billie-cat) [inline src/main/java/com/googleinterns/zoomtube/servlets/DiscussionServlet.java:84]: I mentioned you should make a variable for new Date(0) to indicate it is for current time.

Date currentTime = new Date(0)

then pass that in. This adds clarity as to what it represents. Also, please do the same for new Date(Clock.systemUTC().millis(). I am not sure what this represents.

File: src/main/java/com/googleinterns/zoomtube/servlets/DiscussionServlet.java [hunk 13]
```diff
@@ -62,34 +64,29 @@ public void init() throws ServletException {
...
+    } else {
+      comment = CommentUtil.createEntityWithParent(
+          lecture, parent, new Date(0), author, content, new Date(Clock.systemUTC().millis()));
+    }
+    datastore.put(comment);
```
Comment (billie-cat) [inline src/main/java/com/googleinterns/zoomtube/servlets/DiscussionServlet.java:89]: For simplicity I would do the following:

if (parent == null) {
  datastore.put(CommentUtil.createEntityWithNoParent(...);
  return;
}
datastore.put(CommentUtil.createEntity(...);

File: src/main/java/com/googleinterns/zoomtube/utils/CommentUtil.java [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+    if (parentKey != null) {
+      builder.setParentKey(parentKey);
+    }
+    return builder.build();
+  }
+
+  /**
+   * Creates and returns an entity with the specified properties and no parent comment.
+   */
+  public static Entity createEntityNoParent(
```
Comment (billie-cat) [inline src/main/java/com/googleinterns/zoomtube/utils/CommentUtil.java:62]: nit: I would actually have these classes overload. So I would keep them both with the name createEntity.

Comment (n-wach) [reply src/main/java/com/googleinterns/zoomtube/utils/CommentUtil.java:62]: Changed!

File: src/main/java/com/googleinterns/zoomtube/servlets/DiscussionServlet.java (hunk 12 above)
Comment (n-wach) [reply src/main/java/com/googleinterns/zoomtube/servlets/DiscussionServlet.java:84]: Will clarify.  `new Date(0)` doesn't create a date with the current time.  That's the noargs constructor.  `new Date(0)` creates a date at unix epoch, which we use to represent the start of the video.

Comment (n-wach) [reply src/main/java/com/googleinterns/zoomtube/servlets/DiscussionServlet.java]: Should be clearer now!

File: src/main/java/com/googleinterns/zoomtube/servlets/DiscussionServlet.java (hunk 13 above)
Comment (n-wach) [reply src/main/java/com/googleinterns/zoomtube/servlets/DiscussionServlet.java]: Cleaned this up.  I had to duplicate one line of code but got rid of the else

Comment (billie-cat) [reply src/main/java/com/googleinterns/zoomtube/servlets/DiscussionServlet.java]: I think this response.setStatus was added. Because of that I would just use a if/else statement so this line does not need to be duplicated.

File: src/test/java/com/googleinterns/zoomtube/utils/CommentUtilTest.java [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+    assertThat(comment.parentKey().isPresent()).isTrue();
+    assertThat(comment.parentKey().get()).isEqualTo(parent);
+    assertThat(comment.timestamp()).isEqualTo(timestamp);
+    assertThat(comment.author()).isEqualTo(author);
+    assertThat(comment.content()).isEqualTo(content);
+    assertThat(comment.created()).isEqualTo(dateNow);
+  }
+
+  @Test
+  public void createEntity_shouldReturnEntityWithProperties_noParent() throws IOException {
```
Comment (billie-cat) [inline src/test/java/com/googleinterns/zoomtube/utils/CommentUtilTest.java:77]:
```suggestion
  public void createEntity_noParent_shouldReturnEntityWithProperties() throws IOException {
```

Remember that conditions are in the middle and noParent is a condition.

File: src/test/java/com/googleinterns/zoomtube/utils/CommentUtilTest.java [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+
+    assertThat(entity.getProperty(CommentUtil.LECTURE)).isEqualTo(lectureKey);
+    assertThat(entity.getProperty(CommentUtil.TIMESTAMP)).isEqualTo(timestamp);
+    assertThat(entity.getProperty(CommentUtil.AUTHOR)).isEqualTo(author);
+    assertThat(entity.getProperty(CommentUtil.CONTENT)).isEqualTo(content);
+    assertThat(entity.getProperty(CommentUtil.CREATED)).isEqualTo(dateNow);
+  }
+
+  @Test
+  public void createEntity_shouldReturnEntityWithProperties_withParent() throws IOException {
```
Comment (billie-cat) [inline src/test/java/com/googleinterns/zoomtube/utils/CommentUtilTest.java:94]:
```suggestion
  public void createEntity_withParent_shouldReturnEntityWithProperties() throws IOException {
```

File: src/test/java/com/googleinterns/zoomtube/utils/CommentUtilTest.java (hunk 15 above)
Comment (n-wach) [reply src/test/java/com/googleinterns/zoomtube/utils/CommentUtilTest.java]: Forgot, thanks!

File: src/test/java/com/googleinterns/zoomtube/utils/CommentUtilTest.java (hunk 16 above)
Comment (n-wach) [reply src/test/java/com/googleinterns/zoomtube/utils/CommentUtilTest.java]: Changed
